File: receiverPI_KBC_TCP.cpp
#include <iostream>
#include <cstring>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <chrono>

// Match C++ Structures 
enum CommandType { COORDINATES = 1, GRIPPER = 2 };

#pragma pack(push, 1) // Ensures cross-platform memory alignment
struct CommandPacket {
    int type;
    float x;
    float y;
    float z;
    int gripperAction;
};

struct ResponsePacket {
    char statusMessage[64];
    int limitReached; 
};
#pragma pack(pop)

// Artificial Bounds & State Variables 
float currentX = 0.0f, currentY = 0.0f, currentZ = 0.0f;
const float LIMIT_MIN = -5.0f;
const float LIMIT_MAX = 5.0f;
int currentGripperState = -1;  // -1 = Uninitialized, 0 = Closed, 1 = Open

// Forward Declarations
void processMovement(const CommandPacket& packet, ResponsePacket& res);
void processGripper(const CommandPacket& packet, ResponsePacket& res);

int main() {
    const int port = 8080;
    int server_fd, client_fd;
    struct sockaddr_in address;
    int opt = 1;
    int addrlen = sizeof(address);

    if ((server_fd = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
        std::cerr << "Socket creation failed\n";
        return -1;
    }

    if (setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt))) {
        std::cerr << "setsockopt failed\n";
        return -1;
    }

    address.sin_family = AF_INET;
    address.sin_addr.s_addr = INADDR_ANY; 
    address.sin_port = htons(port);

    if (bind(server_fd, (struct sockaddr*)&address, sizeof(address)) < 0) {
        std::cerr << "Bind failed\n";
        return -1;
    }

    if (listen(server_fd, 3) < 0) {
        std::cerr << "Listen failed\n";
        return -1;
    }

    std::cout << "Server listening on port " << port << "..." << std::endl;

    while (true) {
        client_fd = accept(server_fd, (struct sockaddr*)&address, (socklen_t*)&addrlen);
        if (client_fd < 0) {
            std::cerr << "Accept connection failed\n";
            continue;
        }

        std::cout << "Client Connected.\n";
        bool wasMovingLastTick = false;
        auto lastPacketTime = std::chrono::steady_clock::now();

        while (true) {
            CommandPacket packet;
            ssize_t bytesRead = recv(client_fd, &packet, sizeof(CommandPacket), MSG_DONTWAIT);

            if (bytesRead == sizeof(CommandPacket)) {
                lastPacketTime = std::chrono::steady_clock::now(); 
                ResponsePacket response = {0};

                if (packet.type == COORDINATES) {
                    processMovement(packet, response);
                    wasMovingLastTick = true;
                } 
                else if (packet.type == GRIPPER) {
                    processGripper(packet, response);
                }

                send(client_fd, &response, sizeof(ResponsePacket), 0);
            } 
            else if (bytesRead == 0) {
                break;
            }

            auto currentTime = std::chrono::steady_clock::now();
            auto elapsedTime = std::chrono::duration_cast<std::chrono::milliseconds>(currentTime - lastPacketTime).count();

            if (wasMovingLastTick && (elapsedTime > 90)) {
                ResponsePacket idleResponse = {0};
                std::strcpy(idleResponse.statusMessage, "Status: Idle");
                idleResponse.limitReached = 0;

                send(client_fd, &idleResponse, sizeof(ResponsePacket), 0);
                wasMovingLastTick = false; 
            }

            usleep(5000); 
        }

        close(client_fd);
        std::cout << "Client Disconnected.\n";
    }

    close(server_fd);
    return 0;
}

// Logic Processors

void processMovement(const CommandPacket& packet, ResponsePacket& res) {
    res.limitReached = 0;

    // Evaluate X Axis
    if (packet.x != 0.0f) {
        float nextX = currentX + packet.x;
        if (nextX < LIMIT_MIN) {
            std::strcpy(res.statusMessage, "Limit reached: -X limit reached!");
            res.limitReached = 1;
            // LIMIT REACHED
        } else if (nextX > LIMIT_MAX) {
            std::strcpy(res.statusMessage, "Limit reached: +X limit reached!");
            res.limitReached = 1;
            // LIMIT REACHED
        } else {
            currentX = nextX;
            std::strcpy(res.statusMessage, packet.x > 0 ? "Moving in +X direction" : "Moving in -X direction");
            // MOVEMENT HERE
        }
    }
    // Evaluate Y Axis
    else if (packet.y != 0.0f) {
        float nextY = currentY + packet.y;
        if (nextY < LIMIT_MIN) {
            std::strcpy(res.statusMessage, "Limit reached: -Y limit reached!");
            res.limitReached = 1;
            // LIMIT REACHED
        } else if (nextY > LIMIT_MAX) {
            std::strcpy(res.statusMessage, "Limit reached: +Y limit reached!");
            res.limitReached = 1;
            // LIMIT REACHED
        } else {
            currentY = nextY;
            std::strcpy(res.statusMessage, packet.y > 0 ? "Moving in +Y direction" : "Moving in -Y direction");
            // MOVEMENT HERE
        }
    }
    // Evaluate Z Axis
    else if (packet.z != 0.0f) {
        float nextZ = currentZ + packet.z;
        if (nextZ < LIMIT_MIN) {
            std::strcpy(res.statusMessage, "Limit reached: -Z limit reached (Descend)!");
            res.limitReached = 1;
            // LIMIT REACHED
        } else if (nextZ > LIMIT_MAX) {
            std::strcpy(res.statusMessage, "Limit reached: +Z limit reached (Ascend)!");
            res.limitReached = 1;
            // LIMIT REACHED
        } else {
            currentZ = nextZ;
            std::strcpy(res.statusMessage, packet.z > 0 ? "Moving in +Z direction (Ascend)" : "Moving in -Z direction (Descend)");
            // MOVEMENT HERE
        }
    }
}

void processGripper(const CommandPacket& packet, ResponsePacket& res) {
    res.limitReached = 0;

    if (packet.gripperAction == 1) { // Requesting Open
        if (currentGripperState == 1) {
            std::strcpy(res.statusMessage, "Gripper status: Gripper already open!");
        } else {
            currentGripperState = 1;
            std::strcpy(res.statusMessage, "Gripper action: Opening gripper...");
            // OPEN GRIPPER
        }
    } 
    else if (packet.gripperAction == 0) { // Requesting Close
        if (currentGripperState == 0) {
            std::strcpy(res.statusMessage, "Gripper status: Gripper already closed!");
        } else {
            currentGripperState = 0;
            std::strcpy(res.statusMessage, "Gripper action: Closing gripper...");
            // CLOSE GRIPPER
        }
    }
}

File: receiverPI_KBC_TCP.py
import socket
import time
import struct

# --- NETWORK CONFIGURATION ---
HOST = '127.0.0.1'  # Local loopback for single-device Pi 5 setup
PORT = 9000         # Matched to local sender script port

# --- STRUCT FORMATS (Matching C++ / Sender Packing) ---
# CommandPacket: int32, float, float, float, int32 -> '<ifffi' (20 bytes)
COMMAND_PACKET_FMT = '<ifffi'
COMMAND_PACKET_SIZE = struct.calcsize(COMMAND_PACKET_FMT)

# ResponsePacket: 64-byte char array, int32 -> '<64si' (68 bytes)
RESPONSE_PACKET_FMT = '<64si'

# --- STATE VARIABLES & BOUNDS ---
currentX, currentY, currentZ = 0.0, 0.0, 0.0
LIMIT_MIN = -5.0
LIMIT_MAX = 5.0
currentGripperState = -1  # -1 = Uninitialized, 0 = Closed, 1 = Open


def pad_status_message(msg_bytes: bytes) -> bytes:
    """Safely pads or truncates byte messages to exactly 64 bytes for struct packing."""
    return msg_bytes[:64].ljust(64, b'\x00')


def processMovement(packet_x, packet_y, packet_z):
    global currentX, currentY, currentZ
    limit_reached = 0
    status_message = b""

    # Evaluate X Axis
    if packet_x != 0.0:
        nextX = currentX + packet_x
        if nextX < LIMIT_MIN:
            status_message = b"Limit reached: -X limit reached!"
            limit_reached = 1
            # LIMIT REACHED
        elif nextX > LIMIT_MAX:
            status_message = b"Limit reached: +X limit reached!"
            limit_reached = 1
            # LIMIT REACHED
        else:
            currentX = nextX
            status_message = b"Moving in +X direction" if packet_x > 0 else b"Moving in -X direction"
            # MOVEMENT HERE

    # Evaluate Y Axis
    elif packet_y != 0.0:
        nextY = currentY + packet_y
        if nextY < LIMIT_MIN:
            status_message = b"Limit reached: -Y limit reached!"
            limit_reached = 1
            # LIMIT REACHED
        elif nextY > LIMIT_MAX:
            status_message = b"Limit reached: +Y limit reached!"
            limit_reached = 1
            # LIMIT REACHED
        else:
            currentY = nextY
            status_message = b"Moving in +Y direction" if packet_y > 0 else b"Moving in -Y direction"
            # MOVEMENT HERE

    # Evaluate Z Axis
    elif packet_z != 0.0:
        nextZ = currentZ + packet_z
        if nextZ < LIMIT_MIN:
            status_message = b"Limit reached: -Z limit reached (Descend)!"
            limit_reached = 1
            # LIMIT REACHED
        elif nextZ > LIMIT_MAX:
            status_message = b"Limit reached: +Z limit reached (Ascend)!"
            limit_reached = 1
            # LIMIT REACHED
        else:
            currentZ = nextZ
            status_message = b"Moving in +Z direction (Ascend)" if packet_z > 0 else b"Moving in -Z direction (Descend)"
            # MOVEMENT HERE

    return status_message, limit_reached


def processGripper(gripper_action):
    global currentGripperState
    limit_reached = 0
    status_message = b""

    if gripper_action == 1:  # Requesting Open
        if currentGripperState == 1:
            status_message = b"Gripper status: Gripper already open!"
        else:
            currentGripperState = 1
            status_message = b"Gripper action: Opening gripper..."
            # OPEN GRIPPER
    elif gripper_action == 0:  # Requesting Close
        if currentGripperState == 0:
            status_message = b"Gripper status: Gripper already closed!"
        else:
            currentGripperState = 0
            status_message = b"Gripper action: Closing gripper..."
            # CLOSE GRIPPER

    return status_message, limit_reached


def main():
    server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    server_socket.bind((HOST, PORT))
    server_socket.listen(3)

    print(f"Hardware Mover listening on {HOST}:{PORT}...")

    while True:
        client_socket, client_address = server_socket.accept()
        print(f"Client connected from {client_address}.")

        client_socket.setblocking(False)
        rx_buffer = bytearray()
        wasMovingLastTick = False
        lastPacketTime = time.time()

        while True:
            try:
                # Accumulate data into buffer to handle TCP fragmentation
                data = client_socket.recv(1024)
                if not data:
                    print("Client disconnected gracefully.")
                    break
                rx_buffer.extend(data)

                # Process all fully received command packets
                while len(rx_buffer) >= COMMAND_PACKET_SIZE:
                    packet_bytes = rx_buffer[:COMMAND_PACKET_SIZE]
                    del rx_buffer[:COMMAND_PACKET_SIZE]

                    lastPacketTime = time.time()  # Refresh timeout clock

                    p_type, p_x, p_y, p_z, p_gripper = struct.unpack(COMMAND_PACKET_FMT, packet_bytes)

                    status_message = b""
                    limit_reached = 0

                    if p_type == 1:  # COORDINATES
                        status_message, limit_reached = processMovement(p_x, p_y, p_z)
                        wasMovingLastTick = True
                    elif p_type == 2:  # GRIPPER
                        status_message, limit_reached = processGripper(p_gripper)

                    # Pack and send 68-byte response
                    padded_msg = pad_status_message(status_message)
                    response = struct.pack(RESPONSE_PACKET_FMT, padded_msg, limit_reached)
                    client_socket.sendall(response)

            except BlockingIOError:
                pass  # No socket data available to read right now
            except (ConnectionResetError, BrokenPipeError):
                print("Client disconnected unexpectedly.")
                break

            # Continuous key monitoring (90 ms idle timeout check)
            elapsed_time_ms = (time.time() - lastPacketTime) * 1000
            if wasMovingLastTick and (elapsed_time_ms > 90):
                try:
                    idle_msg = pad_status_message(b"Status: Idle")
                    idle_response = struct.pack(RESPONSE_PACKET_FMT, idle_msg, 0)
                    client_socket.sendall(idle_response)
                except (ConnectionResetError, BrokenPipeError):
                    break
                wasMovingLastTick = False  # Prevent spamming idle message

            time.sleep(0.005)  # 5 ms loop padding

        client_socket.close()


if __name__ == '__main__':
    main()